packages/protocol/contracts/L1/libs/LibL2Tokenomics.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.18;
+
+import {AddressResolver} from "../../common/AddressResolver.sol";
+import {LibMath} from "../../libs/LibMath.sol";
+import {LibFixedPointMath} from "../../thirdparty/LibFixedPointMath.sol";
+import {
+    SafeCastUpgradeable
+} from "@openzeppelin/contracts-upgradeable/utils/math/SafeCastUpgradeable.sol";
+import {TaikoData} from "../TaikoData.sol";
+
+library LibL2Tokenomics {
+    using LibMath for uint256;
+    using SafeCastUpgradeable for uint256;
+
+    error L1_1559_GAS_CHANGE_MISMATCH(uint64 expectedRatio, uint64 actualRatio);
+    error L1_1559_X_SCALE_TOO_LARGE();
+    error L1_1559_Y_SCALE_TOO_LARGE();
+    error L1_OUT_OF_BLOCK_SPACE();
+
+    function getL2Basefee(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        uint32 gasLimit
+    ) internal view returns (uint64 basefee, uint64 newGasExcess) {
+        if (config.gasIssuedPerSecond == 0) {
+            // L2 1559 disabled
+            return (0, 0);
+        }
+
+        unchecked {
+            uint256 reduced = (block.timestamp - state.lastProposedAt) *
+                config.gasIssuedPerSecond;
+            newGasExcess = uint64(reduced.max(state.l2GasExcess) - reduced);
+        }
+
+        basefee = calcL2Basefee({
+            l2GasExcess: newGasExcess,
+            xscale: state.l2Xscale,
+            yscale: uint256(state.l2Yscale) << 64,
+            gasAmount: gasLimit
+        }).toUint64();
+
+        newGasExcess += gasLimit;
+    }
+
+    function calcL2BasefeeParams(
+        uint64 gasExcessMax,
+        uint64 basefeeInitial,
+        uint64 gasTarget,
+        uint64 expected2X1XRatio
+    ) internal pure returns (uint64 l2GasExcess, uint64 xscale, uint64 yscale) {
+        assert(gasExcessMax != 0);
+
+        l2GasExcess = gasExcessMax / 2;
+
+        // calculate xscale
+        uint256 _xscale = LibFixedPointMath.MAX_EXP_INPUT / gasExcessMax;
+        if (_xscale >= type(uint64).max) {
+            revert L1_1559_X_SCALE_TOO_LARGE();
+        }
+        xscale = uint64(_xscale);
+
+        // calculate yscale
+        uint256 _yscale = calcL2Basefee(
+            l2GasExcess,
+            xscale,
+            basefeeInitial,
+            gasTarget
+        );
+        if ((_yscale >> 64) >= type(uint64).max) {
+            revert L1_1559_Y_SCALE_TOO_LARGE();
+        }
+
+        yscale = uint64(_yscale >> 64);
+
+        // Verify the gas price ratio between two blocks, one has
+        // 2*gasTarget gas and the other one has gasTarget gas.
```

Comment:
So only enough precision to guarantee 2x target or 0.5x target? But `l2GasExcess` will have very different values when running the chain, so is this test good enough to ensure a wide range there as well? (maybe [0,  l2GasExcess] is the idea for that?)

Reply:
In theory, the current `l2GasExcess` shouldn't matter for calculating the % difference between basefees for a block of T and and a block 2T - the e^curve has this nice property guaranteed. But we do need a few more tests to test when `l2GasExcess` is 0 and when it is close to the max possible value. More tests will be added.

Reply:
the `test1559_2X1XRatio` test shows that using some reasonable parameters, the ratio remains while gasExcess is in the range of [gasExcessMax/2 - 50 * gasTarget, gasExcessMax/2 + 50 * gasTarget]. When gasExcess is close to 0, the price change ratio will be lower due to exp precision, and when gasExcess is close to the gasExcessMax, the fee is so high that an uint64 cannot hold it, in the code,  we also revert if this happens.